### DISCUSSION TITLE
Why does my gltf look dull?

### DISCUSSION THREAD

**Discussion by dov** [vsg-dev/VulkanSceneGraph#1235] [General]
Hello,

In one of my free software programs, I'm exporting gltf. However when looking at the resulting gltf in vsgviewer, it looks very "dull" and unlit, compared to when loading the same gltf file into blender.

As other gltf files look much better in vsgviewer, I'm sure that I'm "too blame" in my construction of the gltf. Currently, my vertex buffer in the gltf file does not contain any normals, and I'm relying on autogenerated normals. Could that be the problem?

Here is the how the file looks in vsgviewer from a couple of different angles:

![vsg3](https://github.com/vsg-dev/VulkanSceneGraph/assets/338235/0fbdd5db-9785-4643-9e36-9c958e8475b9)
![vsg2](https://github.com/vsg-dev/VulkanSceneGraph/assets/338235/8219ab9f-2bfc-442c-88d7-50e4c87ab2d9)
![vsg1](https://github.com/vsg-dev/VulkanSceneGraph/assets/338235/63f60555-6772-4c56-ac58-2d30fe49d071)

And here is how it looks like in blender:

![bl](https://github.com/vsg-dev/VulkanSceneGraph/assets/338235/7b0e2bb8-2e17-4277-a0ff-c92b0866d877)

And here is the glb file (wrapped in a zip file):

[plate.zip](https://github.com/user-attachments/files/16116937/plate.zip)

So my question what should I change either the gltf file or in (my version of) vsgviewer to get rendering similar to Blender?

**Comment by robertosfield** [comment]
I haven't had chance to look at the model but it does sound like the original model doesn't have normals and the VSG is simply honoring this, while blending is adding them by default.  

In a modelling tools you could load the model, generate the normals then saving to gltf and then load into them in the VSG application.  Or use the vsgXchange::assimp loaders support computing normals for you.  The vsgXchange has a couple of options:

~~~ sh
        vsg::Options::Value      type
        -------------------      ----
        crease_angle             float
        generate_sharp_normals   bool
        generate_smooth_normals  bool
        two_sided                bool
~~~

You can pass this options programatically using:

     options->setValue("generate_sharp_normals", true);
     auto model = vsg::read_cast<vsg::Node>("mymodel.gltf");

Or in vsgviewer:

     vsgviewer --generate_sharp_normals mymodel.gltf

**Reply by dov** [reply]
Thanks @robertosfield . The missing normals together with poor setup of the PBR material was indeed the problem.

**Reply by dov** [reply]
Continuing investing this issue, I found the following note in the glTF specification:

``When normals are not specified, client implementations MUST calculate flat normals and the provided tangents (if present) MUST be ignored.'' [1]

From this I understand that if the gltf file does not contain any normal information, the viewer should display the scene in the same way as if `--generate_sharp_models` had be given. But from looking at plate.gltf in vsgviewer with and without the option, I get different results, so this looks like a bug. Did I miss something?

[1] https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html

**Reply by robertosfield** [reply]
The VSG doesn't directly load gltf files, it uses assimp to load them, and assimp decides what to do, the VSG just converts the graph that assimp creates.  Would be worth having a look at the assimp Issue tracker to see if others had similar problems.
